pydm/widgets/timeplot.py

```diff
@@ -1,160 +1,255 @@
+      return None
+    return self.channel.address
+  
+  @address.setter
+  def address(self, new_address):
+    if new_address is None or len(str(new_address)) < 1:
+      self.channel = None
+      return
+    self.channel = PyDMChannel(address=new_address, connection_slot=self.connectionStateChanged, value_slot=self.receiveNewValue)
+  
+  @property
+  def color_string(self):
+    return str(utilities.colors.svg_color_from_hex(self.color.name(), hex_on_fail=True))
+  
+  @color_string.setter
+  def color_string(self, new_color_string):
+    self.setPen(QColor(str(new_color_string)))
+  
+  @property
+  def color(self):
+    return self.opts['pen'].color()
+    
+  @pyqtSlot(bool)
+  def connectionStateChanged(self, connected):
+    #Maybe change pen stroke?
+    self.connected = connected
+  
+  @pyqtSlot(float)
+  @pyqtSlot(int)
+  def receiveNewValue(self, new_value):
+    if self._update_mode == PyDMTimePlot.SynchronousMode:
+      self.data_buffer = np.roll(self.data_buffer, -1)
+      self.data_buffer[0, self._bufferSize - 1] = time.time()
+      self.data_buffer[1, self._bufferSize - 1] = new_value
+      if self.points_accumulated < self._bufferSize:
+        self.points_accumulated = self.points_accumulated + 1
+    elif self._update_mode == PyDMTimePlot.AsynchronousMode:
+      self.latest_value = new_value
+      
+  @pyqtSlot()
+  def asyncUpdate(self):
+    if self._update_mode != PyDMTimePlot.AsynchronousMode:
+      return
+    self.data_buffer = np.roll(self.data_buffer, -1)
+    self.data_buffer[0, self._bufferSize - 1] = time.time()
+    self.data_buffer[1, self._bufferSize - 1] = self.latest_value
+    if self.points_accumulated < self._bufferSize:
+      self.points_accumulated = self.points_accumulated + 1    
+    
+  def initialize_buffer(self):
+    self.points_accumulated = 0
+    #If you don't specify dtype=float, you don't have enough resolution for the timestamp data.
+    self.data_buffer = np.zeros((2,self._bufferSize), order='f',dtype=float)
+    self.data_buffer[0].fill(time.time())
+    
+  def getBufferSize(self):
+    return int(self._bufferSize)
+  
+  def setBufferSize(self, value):
+    if self._bufferSize != int(value):
+      self._bufferSize = max(int(value),1)
+      self.initialize_buffer()
+
+  def resetBufferSize(self):
+    if self._bufferSize != 1200:
+      self._bufferSize = 1200
+      self.initialize_buffer()
+  
+  @pyqtSlot()
+  def redrawCurve(self):
+    if self.connected:
+      self.setData(y=self.data_buffer[1, -self.points_accumulated:], x=self.data_buffer[0, -self.points_accumulated:])
+  
+  def setUpdatesAsynchronously(self, value):
+    if value == True:
+      self._update_mode = PyDMTimePlot.AsynchronousMode
+    else:
+      self._update_mode = PyDMTimePlot.SynchronousMode
+    self.initialize_buffer()
+
+  def resetUpdatesAsynchronously(self):
+    self._update_mode = PyDMTimePlot.SynchronousMode
+    self.initialize_buffer()
+  
+  def max_x(self):
```

Comment:
I believe that the idea of this method it to return the maximum value of x dimension (time values). It is returning a float. max_x() is used in updateXAxis to find the maximum time value among all curves, so the plot can be adjusted to fit all curves inside.